services.sh

```diff
@@ -214,6 +214,99 @@ function manta_setup_common_log_rotation {
+# (i.e., config-agent and registrar) as well as any additional Manta service
+# logs (i.e., whatever's installed in this zone, like buckets-api, mako, or
+# whatever).  The rotated logs are dropped into /var/log/manta/upload.  cron is
+# configured to run logadm every hour on the hour, so each hour's logs get
+# dropped into /var/log/manta/upload at the top of each hour.  Separately, cron
+# is configured to run the log uploader script (backup.sh) every hour, which
+# uploads any logs it finds in /var/log/manta/upload up to Manta and then
+# removes the local copies.  If Manta is down when this happens, the log files
+# will remain in /var/log/manta/upload until the next hour when Manta is up, at
+# which point they'll be uploaded and then the local copies will be removed.
+#
+# This function relies on SERVICE already being set up in order to determine the
+# number of instances of SERVICE so it SHOULD NOT be called until after SERVICE
+# has been initialized.
+function manta_buckets_setup_common_log_rotation {
+    echo "Setting up common log rotation entries"
+
+    #
+    # Create /var/log/manta/upload, where we store files ready to be uploaded.
+    #
+    mkdir -p /var/log/manta/upload
+    chown root:sys /var/log/manta/upload
+
+    #
+    # Copy the log uploader into a place where cron can run it.
+    #
+    local DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+    mkdir -p /opt/smartdc/common/sbin
+    cp ${DIR}/backup.sh /opt/smartdc/common/sbin
+    chmod 755 /opt/smartdc/common/sbin/backup.sh
+    cp ${DIR}/logrotateandupload.sh /opt/smartdc/common/sbin
+    chmod 755 /opt/smartdc/common/sbin/logrotateandupload.sh
+    chown root:sys /opt/smartdc/common
+
+    #
+    # Ensure that log rotation HUPs *r*syslog.
+    #
+    logadm -r /var/adm/messages
+    logadm -w /var/adm/messages -C 4 -a 'kill -HUP `cat /var/run/rsyslogd.pid`'
+
+    #
+    # We want to rotate smf_logs and /var/log/*.log last so they don't override
+    # our log rotation rules for files otherwise caught in their glob pattern,
+    # We'll remove those entries here and re-add them below, in
+    # manta_common_setup_end, after all of our other changes.
+    #
+    # The "/var/log/*.debug" entry is vestigial for old vmadm logs (see
+    # MANTA-3684) and can be removed.
+    #
+    logadm -r smf_logs
+    logadm -r '/var/log/*.log'
+    logadm -r '/var/log/*.debug'
+
+    #
+    # Add the logadm configurations for the config-agent and registrar services
+    # (which are present in every zone), plus the one we've been given.
+    #
+    manta_add_logadm_entry "config-agent"
+    manta_add_logadm_entry "registrar"
+    if [[ $# -ge 1 ]]; then
+        for port in `svcs -H -o fmri $1 | grep -v default | sed s/'.*-'//`
+        do
+            pattern="$logdir/*:$service-$port.log"
+            logadm -w "$1-$port" -C 48 -c -p 1h \
+                   -t "/var/log/manta/upload/$1_\$nodename_%FT%H:00:00_$port.log" \
```

Comment:
And would you also like to be a little bit of a hero and change to *not* having colons in the rotated log file names? It would take this patch:

```diff
--- a/logrotateandupload.sh
+++ b/logrotateandupload.sh
@@ -175,9 +175,9 @@ sleep $(($sleeptime * 60))


 # Files look like this:
-#     buckets-api_0db94777-555d-4f1a-a87f-b1e2ee13c025_2012-10-17T21:00:00_8081.log
+#     buckets-api_0db94777-555d-4f1a-a87f-b1e2ee13c025_20121017T210000_8081.log
 # or this:
-#     boray_b3a7f519-1096-4e47-9a56-efbd1ab8b692_2012-10-17T21:00:00.log
+#     boray_b3a7f519-1096-4e47-9a56-efbd1ab8b692_20121017T210000.log
 # And we transform them to this in manta:
 #     /poseidon/stor/logs/buckets-api/2012/10/17/20/0db94777.8081.log
 #     /poseidon/stor/logs/boray/2012/10/17/20/b3a7f519.log
@@ -200,7 +200,8 @@ do
     if [[ ! -z "$instance" ]]; then
         instance=".$instance"
     fi
-    time=$(date -d \@$(( $(date -d $logtime "+%s") - 3600 )) "+%Y/%m/%d/%H")
+    isotime=$(echo $logtime | sed -E 's/^(....)(..)(..)T(..)(..)(..)$/\1-\2-\3T\4:\5:\6/')
+    time=$(date -d \@$(( $(date -d $isotime "+%s") - 3600 )) "+%Y/%m/%d/%H")
     key="/logs/$service/$time/$zone$instance.log"
     mkdirp $service $time
     manta_put "$key" "$LOG_TYPE" "-T $f"
diff --git a/services.sh b/services.sh
index 1e5dc37..45a746f 100644
--- a/services.sh
+++ b/services.sh
@@ -286,7 +286,7 @@ function manta_buckets_setup_common_log_rotation {
         do
             pattern="$logdir/*:$service-$port.log"
             logadm -w "$1-$port" -C 48 -c -p 1h \
-                   -t "/var/log/manta/upload/$1_\$nodename_%FT%H:00:00_$port.log" \
+                   -t "/var/log/manta/upload/$1_\$nodename_%Y%m%dT%H%M%S_$port.log" \
                    "$pattern" || fatal "unable to create logadm entry"
         done
```

I could perhaps do this as a separate change just after yours, if you like. The argument *for* doing this is perhaps a little stronger for the Triton case where hourly rotated logs actually still stick around in the zone for local debugging for a number of hours, even after being uploaded.